renku/service/controllers/graph_build.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku graph build controller."""
+import os
+
+from renku.core.commands.migrate import migrations_check
+from renku.core.errors import RenkuException
+from renku.service.controllers.api.abstract import ServiceCtrl
+from renku.service.controllers.api.mixins import ReadOperationMixin
+from renku.service.jobs.contexts import enqueue_retry
+from renku.service.jobs.graph import graph_build_job
+from renku.service.jobs.queues import GRAPH_JOB_QUEUE
+from renku.service.serializers.graph import GraphBuildRequest, GraphBuildResponseRPC
+from renku.service.views import result_response
+
+
+class GraphBuildCtrl(ServiceCtrl, ReadOperationMixin):
+    """Controller for build graph endpoint."""
+
+    REQUEST_SERIALIZER = GraphBuildRequest()
+    RESPONSE_SERIALIZER = GraphBuildResponseRPC()
+
+    def __init__(self, identity, request_data):
+        """Construct a datasets list controller."""
+        self.identity = identity
+        self.ctx = GraphBuildCtrl.REQUEST_SERIALIZER.load(request_data)
+        super(GraphBuildCtrl, self).__init__(None, identity, request_data)
+
+    @property
+    def context(self):
+        """Controller operation context."""
+        return self.ctx
+
+    def renku_op(self):
+        """Renku operation for the controller."""
+        result = migrations_check().build().execute().output
+
+        if not result[1]:
+            raise RenkuException("project not supported")
+
+        with enqueue_retry(GRAPH_JOB_QUEUE) as queue:
+            queue.enqueue(
+                graph_build_job,
+                self.ctx.get("revision"),
```

Comment:
`--revision` doesn't actually do what it's name implies and I consider is useless at best.

Firstly, it takes a single revision or a range `a..b`, but in the case of a single revision `rev` if interprets it as `rev..HEAD`, which isn't what we want in 99% of the cases.

Then, it doesn't actually check out that revision and generate triples for it, instead if looks at the paths that were changed in the range mentioned above, and then amounts to calling `renku log` with those paths supplied for the `paths` argument, BUT at the HEAD commit. So if the commit edited e.g. `.renku/datasets/uuid/metadata.yaml`, but the dataset was changed after, it will generate triples for that dataset, but with the state it is currently in.

That's why currently the KG checks out each commit and calls `renku log` instead of just checking out the HEAD and calling `renku log --revision`. If we want to support that usecase, we need to use a different approach here.

Also I think we should just get rid of `--revision` altogether, it causes way more problems than it's worth.

Reply:
Thanks for the extended explanation, I agree. But this specific line is just a request parameter with which client specifies which branch/commit should be checkout and graph build of. See comment bellow.